posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc

```diff
@@ -0,0 +1,158 @@
+even though the first 99,000 entities were indexed just fine...
+
+With the JSR-352 mass indexing job, the job will simply resume from the last checkpoint,
+probably somewhere around the 95,000th entity.
+
+There are other advantages (such as being able to pool resources between multiple, unrelated background jobs),
+but this one really is why we think JSR-352 makes sense.
+
+== How to use the JSR-352 mass indexing job?
+
+=== Setup
+
+Within Wildfly https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-configuration-deploy-on-wildfly[use our WildFly modules],
+configure Hibernate ORM and Hibernate Search. Also add a dependency to the `org.hibernate.search.jsr352` module.
+That's all: you can then retrieve a `javax.batch.operations.JobOperator` instance
+in your application through CDI injection.
+
+In a Java SE application, you can use the reference implementation for JSR-352,
+https://github.com/WASdev/standards.jsr352.jbatch#readme[JBatch], in Java SE mode.
+You will need to:
+
+* add a dependency to the following artifacts in your project:
+** `org.hibernate:hibernate-search-jsr352-core:5.9.0.Beta1`
+** `com.ibm.jbatch:com.ibm.jbatch-runtime:1.0`
+** `javax.batch:javax.batch-api:1.0`
+** `javax.inject:javax.inject:1`
+** `org.apache.derby:derby:10.13.1.1`
+* add https://github.com/hibernate/hibernate-search/tree/master/integrationtest/jsr352/src/test/resources/META-INF/services[configuration files]
+  to you classpath under `META-INF/services`,
+* retrieve the `JobOperator` in your application by calling `javax.batch.runtime.BatchRuntime.getJobOperator()`
+(be sure to store the result, as each call may create a new operator).
+* make sure to initialize Hibernate ORM before you try to launch the mass indexing job,
+  and to only shut it down after the last mass indexing job stopped.
+
+As for other JSR-352 runtimes, the mass indexer may or may not work.
+We are in the process of testing at least one other JEE application server and Spring Batch:
+see <<limitations-and-next>> below.
+
+=== Usage
+
+All the examples below assume the variable `jobOperator` contains
+the job operator retrieved as instructed above.
+
+To launch a mass indexing job:
+
+```
+java.util.Properties parameters = org.hibernate.search.jsr352.massindexing.MassIndexingJob.parameters()
+		.forEntities( MyFirstEntity.class, MySecondEntity.class )
+		.build();
+long executionId = jobOperator.start(
+		org.hibernate.search.jsr352.massindexing.MassIndexingJob.NAME,
+		parameters
+		);
+```
+
+You can then query the execution status through this call:
+
+```
+javax.batch.runtime.JobExecution jobExecution = jobOperator.getJobExecution( executionId );
+```
+
+You can suspend the job this way:
+
+```
+long executionId2 = jobOperator.stop( executionId );
+```
+
+To resume the job after suspension or upon failure, you can request a restart this way:
+
+```
+long executionId2 = jobOperator.restart( executionId, parameters );
```

Comment:
Is there a way to force a full restart? Use case: it's running, then I realize it's not using the analyzer configurations I want, to I kill it with the intent to have it start over.

Reply:
You just start another execution (`.start`) with the same parameters. But I'm not sure it would be useful to mention it here... ?

Reply:
sure, np. Was just a thing which came to mind, wanted to make sure that's possible.